rust_src/src/buffers.rs

```diff
@@ -1982,4 +2030,268 @@ pub fn byte_char_debug_check(b: LispBufferRef, charpos: isize, bytepos: isize) {
     }
 }
 
+/// Kill the buffer specified by BUFFER-OR-NAME.
+/// The argument may be a buffer or the name of an existing buffer.
+/// Argument nil or omitted means kill the current buffer.  Return t if the
+/// buffer is actually killed, nil otherwise.
+///
+/// The functions in `kill-buffer-query-functions' are called with the
+/// buffer to be killed as the current buffer.  If any of them returns nil,
+/// the buffer is not killed.  The hook `kill-buffer-hook' is run before the
+/// buffer is actually killed.  The buffer being killed will be current
+/// while the hook is running.  Functions called by any of these hooks are
+/// supposed to not change the current buffer.
+///
+/// Any processes that have this buffer as the `process-buffer' are killed
+/// with SIGHUP.  This function calls `replace-buffer-in-windows' for
+/// cleaning up all windows currently displaying the buffer to be killed.
+#[lisp_fn(min = "0", intspec = "bKill buffer: ")]
+pub fn kill_buffer(buffer_or_name: Option<LispBufferOrName>) -> bool {
+    let mut b = match buffer_or_name {
+        Some(buf_or_name) => LispBufferRef::from(buf_or_name),
+        None => ThreadState::current_buffer_unchecked(),
+    };
+
+    // Avoid trouble for buffers that are already dead.
+    if !b.is_live() || unsafe { thread_check_current_buffer(b.as_mut()) } {
+        return false;
+    }
+
+    // Run hooks with the buffer to be killed the current buffer.
+    let count = unsafe { SPECPDL_INDEX() };
+
+    unsafe {
+        record_unwind_protect(Some(save_excursion_restore), save_excursion_save());
```

Comment:
Out of scope for the PR, but long term I would love to see this protection wrapped in an object so the code could become something like:
```
let unwind = UnwindProtection::start();
.......
if some_reason:
    return unwind.complete();

// finish the function
return blah; // this will also call unwind.complete automatically.
```

Reply:
Alongside the added safety benefit it would also make it more mentally intuitive

Reply:
Perhaps a similar thing could be done with input locking as well, i.e. also having a `InputLock::start()` and `input_lock.release()`